feat: Added optional icon functionality to Button component

luanvdw opened the pull request:
Hi @madebysid 

Herewith a PR in response to #80 

The Button component has been updated such that it can accept an <Icon /> component in the following manner

`startIcon`: which will render an Icon before the children
`endIcon`: which will render an Icon after the children

I have also attached screenshot(s) for reference.

![story](https://user-images.githubusercontent.com/6928171/116128506-50d97d00-a6c9-11eb-8af1-faf4f5f04f9f.png)

![story](https://user-images.githubusercontent.com/6928171/116128526-546d0400-a6c9-11eb-8410-b867499a05c8.gif)

Happy to make updates if needed 👍

madebysid commented on src/components/button/Button.tsx:
```diff
@@ -18,6 +18,10 @@ export type ButtonProps = React.PropsWithChildren<{
+  /** Icon element placed before the children. */
+  startIcon?: React.ReactNode | null
+  /** Icon element placed after the children. */
+  endIcon?: React.ReactNode | null
```
So in general, we try to avoid receiving a React node as a prop in components. This is because it opens up the door to misuse (not malicious, of course).
We want the design system to be very tightly defined, so that you can only do one thing: the thing that was intended.

Following the same logic, you'll see that most components also do not expose a `className` prop to stop consumers from modifying the component. (There are some exceptions, but they;re supposed to be rare)

In this case, instead of receiving a ReactNode, try and receive the `name` of the icon and render the `Icon` yourself inside this component.
In this specific scenario, it forces the consumer to only render icons, and only a "predefined" icon of a specific size. This might seem backwards, but like I said, it is by design :)

madebysid replied:
I also like that you added props to render an icon on either side of the text! That'll definitely come in handy!

madebysid commented on src/components/button/Button.tsx:
```diff
@@ -90,7 +96,13 @@ export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
           )}
         >
+          {startIcon && (
+            <span className="mr-1">{startIcon}</span>
```
Here, try and use the `Icon` component directly like I said in my other comment. You probably won't need a span around it either, since an `Icon` is one of the few components that accepts a `className` prop.

madebysid commented:
Hey @luanvdw, thanks for creating this PR! I've left a few comments that would be awesome to implement.

To be honest, we weren't quite expecting community contributions, so there's no way you could have expected some of these comments, I apologize for that.

Nevertheless, I still do appreciate it, and I'll see if can get a contributing guide up soon :)

madebysid commented:
Ooh it also looks like something's up with the package-lock on your branch (conflict aside). I use `npm 7`'s package-lock format, that's probably what that is!

luanvdw commented:
Hey @madebysid 

Thank you for your detailed feedback on reviewing my Pull Request, especially around the use of ReactNode and how limiting its use is by design 👌 

I have updated the PR with these changes;

1 - The Icon now renders inside the Button component based on the name supplied. I am not certain what your preference is regarding the format of this but since it is similar to the `variant` (in determining the type of icon to be rendered) I implemented it in a similar manner for consistency. This list can grow as more icons are added.

2 - I have also added an iconSize prop which defaults to `"sm"`

3 - I have updated the custom iconArgType for the storybook interaction, such that a user can also select to not have an icon (a null value) to accommodate not having a null value stipulated as an icon option on the actual button component taking into consideration that it is an optional prop. Feel free to let me know if I am not making sense 😅 

Herewith a video of the updated interaction.

![using](https://user-images.githubusercontent.com/6928171/117621522-fbde4200-b171-11eb-85a7-a4c7e9c5cb42.gif)

![storybook](https://user-images.githubusercontent.com/6928171/117621570-04cf1380-b172-11eb-855a-77b001af5f61.gif)

Thanks again for your review and I am happy to make further updates if needed 👍

madebysid commented:
Hey Luan, just wanted to drop in and say thanks for implementing these changes! I'm a little shorthanded right now, but I will review and merge this in as soon as possible!

madebysid commented:
LGTM!

(Also welcome to the team! Excited to work together!)
